runtime/storage.go

```diff
@@ -96,279 +101,149 @@ func NewStorage(ledger atree.Ledger, memoryGauge common.MemoryGauge) *Storage {
-	//
-	// Migration happens in Commit() if unmigrated account has write ops.
-	// NOTE: Commit() is not called by this function.
-	//
-	// Specifically,
-	// - unmigrated account is migrated in Commit() if there are write ops.
-	//   For example, inserting values in unmigrated account triggers migration in Commit().
-	// - unmigrated account is unchanged if there are only read ops.
-	//   For example, iterating values in unmigrated account doesn't trigger migration,
-	//   and checking if domain exists doesn't trigger migration.
-
 	// Get cached domain storage map if it exists.
 
 	domainStorageKey := interpreter.NewStorageDomainKey(s.memoryGauge, address, domain)
 
-	if domainStorageMap := s.cachedDomainStorageMaps[domainStorageKey]; domainStorageMap != nil {
-		return domainStorageMap
-	}
-
-	// Get (or create) domain storage map from existing account storage map
-	// if account is migrated account.
-
-	accountStorageKey := interpreter.NewStorageKey(s.memoryGauge, address, AccountStorageKey)
-
-	accountStorageMap, err := s.getAccountStorageMap(accountStorageKey)
-	if err != nil {
-		panic(err)
-	}
-
-	if accountStorageMap != nil {
-		// This is migrated account.
-
-		// Get (or create) domain storage map from account storage map.
-		domainStorageMap := accountStorageMap.GetDomain(s.memoryGauge, inter, domain, createIfNotExists)
-
-		// Cache domain storage map
+	if s.cachedDomainStorageMaps != nil {
+		domainStorageMap = s.cachedDomainStorageMaps[domainStorageKey]
 		if domainStorageMap != nil {
-			s.cachedDomainStorageMaps[domainStorageKey] = domainStorageMap
+			return domainStorageMap
 		}
-
-		return domainStorageMap
-	}
-
-	// At this point, account is either new or unmigrated account.
-
-	domainStorageMap, err := getDomainStorageMapFromLegacyDomainRegister(s.Ledger, s.PersistentSlabStorage, address, domain)
-	if err != nil {
-		panic(err)
 	}
 
-	if domainStorageMap != nil {
-		// This is a unmigrated account with given domain register.
-
+	defer func() {
 		// Cache domain storage map
-		s.cachedDomainStorageMaps[domainStorageKey] = domainStorageMap
-
-		// Add account to unmigrated account list
-		s.addUnmigratedAccount(address)
-
-		return domainStorageMap
-	}
-
-	// At this point, account is either new account or unmigrated account without given domain.
-
-	// Domain doesn't exist.  Return early if createIfNotExists is false.
-
-	if !createIfNotExists {
-		return nil
-	}
-
-	// Handle unmigrated account
-	unmigrated, err := s.isUnmigratedAccount(address)
-	if err != nil {
-		panic(err)
-	}
-	if unmigrated {
-		// Add account to unmigrated account list
-		s.addUnmigratedAccount(address)
+		if domainStorageMap != nil {
+			s.cacheDomainStorageMap(
+				domainStorageKey,
+				domainStorageMap,
+			)
+		}
+	}()
 
-		// Create new domain storage map
-		domainStorageMap := interpreter.NewDomainStorageMap(s.memoryGauge, s, atree.Address(address))
+	if !s.Config.StorageFormatV2Enabled || s.IsV1Account(address) {
```

Comment:
Hmm, what do you think is the best way to resolve that?

Reply:
Here is one way to reduce register reads.

When feature is enabled,
- for V2 account, "stored" register is read.
- for V1 account and new account,
  - if `createIfNotExists` is false, "stored" register and given domain register are read.  
  - if `createIfNotExists` is true, "stored" register and given domain registers are read first.  If domain doesn't exist, more domain registers are read.

This code can probably be simplified more but it reduces register reads.
 
```Go
if !s.Config.StorageFormatV2Enabled {
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap 
}

if s.IsV2Account(address) {
    domainStorageMap = s.AccountStorageV2.GetDomainStorageMap(...)
    ...
    return domainStorageMap
}
 
if s.HasDomainRegister(domain)
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap
} 

if !createIfNotExists {
   return nil
}

if s.IsNewAccount(address) {
    domainStorageMap = s.AccountStorageV2.GetDomainStorageMap(...)
    ...
    return domainStorageMap
} else {
    domainStorageMap = s.AccountStorageV1.GetDomainStorageMap(...)
    ...
    return domainStorageMap
}
```

Reply:
It would be great to avoid unnecessary reads, even if we have to make the code a bit more complex 👍 
A few questions:
- What is `IsV2Account`? Does it only check for the V2 `stored` register? (Currently we only have `IsV1Account`. Note that I had previously optimized `IsV1Account` to check for the V2 `stored` register first)
- What is `HasDomainRegister`? Is it maybe `hasAccountStorageMap`?
- What is `IsNewAccount`? Would this determine if the account has any register? (e.g. V2 `stored` account storage map register or any of the known V1 domain storage map registers)

Also happy to have a quick sync about this, and please also feel free to open up a follow-up PR 🙏

Reply:
> Also happy to have a quick sync about this, and please also feel free to open up a follow-up PR 🙏

OK, happy to do that.  I will open a follow-up PR and then we can discuss it.

Reply:
@turbolent I think we can merge this PR to feature branch.  

I opened PR #3683 to reduce register reads. The new PR can be reviewed and merged separately.  PTAL :pray:

Reply:
Thank you for opening the PR, I will try to have a look tomorrow